com.unity.ml-agents/Runtime/Sensors/Reflection/ObservableAttribute.cs

```diff
@@ -0,0 +1,296 @@
+        {
+            var sensorsOut = new List<ISensor>();
+            foreach (var(field, attr) in GetObservableFields(o, excludeInherited))
+            {
+                var sensor = CreateReflectionSensor(o, field, null, attr);
+                if (sensor != null)
+                {
+                    sensorsOut.Add(sensor);
+                }
+            }
+
+            foreach (var(prop, attr) in GetObservableProperties(o, excludeInherited))
+            {
+                if (!prop.CanRead)
+                {
+                    // Skip unreadable properties.
+                    continue;
+                }
+                var sensor = CreateReflectionSensor(o, null, prop, attr);
+                if (sensor != null)
+                {
+                    sensorsOut.Add(sensor);
+                }
+            }
+
+            return sensorsOut;
+        }
+
+        /// <summary>
+        /// Create the ISensor for either the field or property on the provided object.
+        /// If the data type is unsupported, or the property is write-only, returns null.
+        /// </summary>
+        /// <param name="o"></param>
+        /// <param name="fieldInfo"></param>
+        /// <param name="propertyInfo"></param>
+        /// <param name="observableAttribute"></param>
+        /// <returns></returns>
+        /// <exception cref="UnityAgentsException"></exception>
+        static ISensor CreateReflectionSensor(object o, FieldInfo fieldInfo, PropertyInfo propertyInfo, ObservableAttribute observableAttribute)
+        {
+            string memberName;
+            string declaringTypeName;
+            Type memberType;
+            if (fieldInfo != null)
+            {
+                declaringTypeName = fieldInfo.DeclaringType.Name;
+                memberName = fieldInfo.Name;
+                memberType = fieldInfo.FieldType;
+            }
+            else
+            {
+                declaringTypeName = propertyInfo.DeclaringType.Name;
+                memberName = propertyInfo.Name;
+                memberType = propertyInfo.PropertyType;
+            }
+
+            string sensorName;
+            if (string.IsNullOrEmpty(observableAttribute.m_Name))
+            {
+                sensorName = $"ObservableAttribute:{declaringTypeName}.{memberName}";
+            }
+            else
+            {
+                sensorName = observableAttribute.m_Name;
+            }
+
+            var reflectionSensorInfo = new ReflectionSensorInfo
+            {
+                Object = o,
+                FieldInfo = fieldInfo,
+                PropertyInfo = propertyInfo,
+                ObservableAttribute = observableAttribute,
+                SensorName = sensorName
+            };
+
+            ISensor sensor = null;
+            if (memberType == typeof(Int32))
+            {
+                sensor = new IntReflectionSensor(reflectionSensorInfo);
+            }
+            if (memberType == typeof(float))
```

Comment:
any reason it's not `else if` all the way down?

Reply:
Doh. It used to be returns so I removed the `else if`s (because Rider complained they were redundant) and I didn't add them back in.

Reply:
or even like how you have a map for type sizes, perhaps that pattern works here as well.  An internal struct maybe that holds `<member_type, internal_sensor_info>` where internalSensorInfo looks like:

```
internal struct SensorInfo 
{
    uint observationSize;
    Class sensorClass;
}
```

not a show stopper, but this seems a bit verbose.

Reply:
Fixed up the `else if`s for now. I'll revisit the others later.